crates/sui-types/src/temporary_store.rs

```diff
@@ -689,6 +688,154 @@ impl<S> TemporaryStore<S> {
     }
 }
 
+impl<S: ObjectStore> TemporaryStore<S> {
+    /// returns lists of (objects whose owner we must authenticate, objects whose owner has already been authenticated)
+    fn get_objects_to_authenticate(
+        &self,
+        sender: &SuiAddress,
+        gas: &[ObjectRef],
+    ) -> SuiResult<(HashSet<ObjectID>, HashSet<ObjectID>)> {
+        let gas_objs: HashSet<&ObjectID> = gas.iter().map(|g| &g.0).collect();
+        let mut objs_to_authenticate = HashSet::new();
+        let mut authenticated_objs = HashSet::new();
+        for (id, obj) in &self.input_objects {
+            if gas_objs.contains(id) {
+                // gas could be owned by either the sender (common case) or sponsor (if this is a sponsored tx,
+                // which we do not know inside this function).
+                // either way, no object ownership chain should be rooted in a gas object
+                // thus, consider object authenticated, but don't add it to authenticated_objs
+                continue;
+            }
+            match &obj.owner {
+                Owner::AddressOwner(a) => {
+                    assert!(sender == a, "Input object not owned by sender");
+                    authenticated_objs.insert(*id);
+                }
+                Owner::Shared { .. } => {
+                    authenticated_objs.insert(*id);
+                }
+                Owner::Immutable => {
+                    // object is authenticated, but it cannot own other objects,
+                    // so we should not add it to `authenticated_objs`
+                }
+                Owner::ObjectOwner(_parent) => {
+                    unreachable!("Input objects must be address owned, shared, or immutable")
+                }
+            }
+        }
+
+        for (id, (_new_obj, kind)) in &self.written {
+            if authenticated_objs.contains(id) || gas_objs.contains(id) {
+                continue;
+            }
+            match kind {
+                WriteKind::Mutate => {
+                    // get owner at beginning of tx, since that's what we have to authenticate against
+                    // _new_obj.owner is not relevant here
+                    let old_obj = self.store.get_object(id)?.unwrap();
+                    match &old_obj.owner {
+                        Owner::ObjectOwner(_parent) => {
+                            objs_to_authenticate.insert(*id);
+                        }
+                        Owner::AddressOwner(_) | Owner::Shared { .. } => {
+                            unreachable!("Should already be in authenticated_objs")
+                        }
+                        Owner::Immutable => {
+                            unreachable!("Immutable objects cannot be written")
```

Comment:
Also a good point!

This should also run on system transactions--it's only disable for dev inspect. Let me try one of those tests and be sure this check is running. Should be easy to special-case the framework here.

Reply:
Perhaps we could also consider adding another kind in WriteKind that's WriteKind::PackageUpgrade? This may make things easier to reason about?

Reply:
Now that you mention it, I think immutable object creation (e.g., package publishing, freezing objects) will also hit this...

Reply:
That would work today,  but we had talked about generally making it possible for system transactions to mutate "immutable" objects (e.g. on checkpoints or epoch boundaries) so that we could disseminate information to single owner transactions.

What we are doing for system packages is a special case of that, but it is not special -- technically we could safely modify any immutable object on the epoch boundary, right?